src/ffi/cpython/abstract_.rs

```diff
@@ -0,0 +1,304 @@
+    pub fn PyObject_VectorcallDict(
+        callable: *mut PyObject,
+        args: *const *mut PyObject,
+        nargsf: usize,
+        kwargs: *mut PyObject,
+    ) -> *mut PyObject;
+
+    #[cfg(all(Py_3_8, not(PyPy)))]
+    #[cfg_attr(not(Py_3_9), link_name = "_PyVectorcall_Call")]
+    pub fn PyVectorcall_Call(
+        callable: *mut PyObject,
+        tuple: *mut PyObject,
+        dict: *mut PyObject,
+    ) -> *mut PyObject;
+}
+
+#[cfg(all(Py_3_8, not(PyPy)))]
+#[inline(always)]
+pub unsafe fn _PyObject_FastCallTstate(
+    tstate: *mut PyThreadState,
+    func: *mut PyObject,
+    args: *const *mut PyObject,
+    nargs: Py_ssize_t,
+) -> *mut PyObject {
+    _PyObject_VectorcallTstate(tstate, func, args, nargs as usize, std::ptr::null_mut())
+}
+
+#[cfg(all(Py_3_8, not(PyPy)))]
+#[inline(always)]
+pub unsafe fn _PyObject_FastCall(
+    func: *mut PyObject,
+    args: *const *mut PyObject,
+    nargs: Py_ssize_t,
+) -> *mut PyObject {
+    _PyObject_FastCallTstate(PyThreadState_GET(), func, args, nargs)
+}
+
+#[cfg(all(Py_3_8, not(PyPy)))]
+#[inline(always)]
+pub unsafe fn _PyObject_CallNoArg(func: *mut PyObject) -> *mut PyObject {
+    _PyObject_VectorcallTstate(
+        PyThreadState_GET(),
+        func,
+        std::ptr::null_mut(),
+        0,
+        std::ptr::null_mut(),
+    )
+}
+
+#[cfg(all(Py_3_8, not(PyPy)))]
+#[inline(always)]
+pub unsafe fn PyObject_CallOneArg(func: *mut PyObject, arg: *mut PyObject) -> *mut PyObject {
+    assert!(!arg.is_null());
+    let _args = [std::ptr::null_mut(), arg];
+    let args = _args.as_ptr().offset(1); // For PY_VECTORCALL_ARGUMENTS_OFFSET
+    let tstate = PyThreadState_GET();
+    let nargsf = 1 | PY_VECTORCALL_ARGUMENTS_OFFSET;
+    return _PyObject_VectorcallTstate(tstate, func, args, nargsf as usize, std::ptr::null_mut());
+}
+
+extern "C" {
+    #[cfg(all(Py_3_9, not(PyPy)))]
+    pub fn PyObject_VectorcallMethod(
+        name: *mut PyObject,
+        args: *const *mut PyObject,
+        nargsf: usize,
+        kwnames: *mut PyObject,
+    ) -> *mut PyObject;
+}
+
+#[cfg(all(Py_3_9, not(PyPy)))]
+#[inline(always)]
+pub unsafe fn PyObject_CallMethodNoArgs(
+    self_: *mut PyObject,
+    name: *mut PyObject,
+) -> *mut PyObject {
+    PyObject_VectorcallMethod(
+        name,
+        &self_,
+        1 | PY_VECTORCALL_ARGUMENTS_OFFSET as usize,
+        std::ptr::null_mut(),
+    )
+}
+
+#[cfg(all(Py_3_9, not(PyPy)))]
+#[inline(always)]
+pub unsafe fn PyObject_CallMethodOneArg(
+    self_: *mut PyObject,
+    name: *mut PyObject,
+    arg: *mut PyObject,
+) -> *mut PyObject {
+    let args = [self_, arg];
+    assert!(!arg.is_null());
+    PyObject_VectorcallMethod(
+        name,
+        args.as_ptr(),
+        2 | PY_VECTORCALL_ARGUMENTS_OFFSET as usize,
+        std::ptr::null_mut(),
+    )
+}
+
+// skipped _PyObject_VectorcallMethodId
+// skipped _PyObject_CallMethodIdNoArgs
+// skipped _PyObject_CallMethodIdOneArg
+
+// skipped _PyObject_HasLen
```

Comment:
Just curious, why are they skipped?

Reply:
I didn't consider them relevant to my PR; I could add them but I'd also quite like to keep as few unlimited API definitions as possible (as they're more likely for us to be work to maintain).

See also #1289 which talks a bit about what I'm thinking.